File: 백준/Gold/2493. 탑/탑.py
import sys
input = sys.stdin.readline

N = int(input())

tops = list(map(int, input().split()))

stack = []
receive = [0] * N

for idx, height in enumerate(tops):
    
    while stack and stack[-1][1] <= height:
        stack.pop()

    if stack:
        receive[idx] = stack[-1][0] + 1

    stack.append((idx, height))

for i in range(N):
    print(receive[i], end=" ")

File: 백준/Gold/2493. 탑/탑.swift
import Foundation

let n = Int(readLine()!)!

var result = ""

var stack: [(height: Int, pos: Int)] = []

let input = readLine()!.split(separator: " ").map { Int(String($0))! }

for i in 0..<input.count {
  
  let value = (height: input[i], pos: i+1)

  while let last = stack.last {

    if last.height >= value.height {
      result += "\(last.pos) "
      stack.append(value)
      break
    } else {
      _ = stack.popLast()
    }
  }

  if stack.isEmpty { 
    result += "0 " 
    stack.append(value)
  }
}


print(result)